maint/qmpi_binding.py

```diff
@@ -265,38 +266,84 @@ def generate_code(parseset: Mapping[str, Any], postfix: str, kind_map: Mapping[s
 qmpi_h_file.write("#define QMPI_MAX_TOOL_NAME_LENGTH 256\n\n");
 qmpi_h_file.write("int QMPI_Register_tool_name(const char *tool_name, void (* init_function_ptr)(int tool_id)) MPICH_API_PUBLIC;\n");
 qmpi_h_file.write("int QMPI_Register_tool_storage(int tool_id, void *tool_storage) MPICH_API_PUBLIC;\n");
 qmpi_h_file.write("int QMPI_Register_function(int calling_tool_id, enum QMPI_Functions_enum function_enum,\n"
                   "                           void (* function_ptr)(void)) MPICH_API_PUBLIC;\n");
 qmpi_h_file.write("\n");
-qmpi_h_file.write("#include <stddef.h>\n"
+qmpi_h_file.write("#include <execinfo.h>\n#include <stddef.h>\n#include <stdio.h>\n#include <stdlib.h>\n#include <string.h>\n"
         "extern MPICH_API_PUBLIC void (**MPIR_QMPI_pointers) (void);\n"
         "extern MPICH_API_PUBLIC void **MPIR_QMPI_storage;\n"
+        "#define BACKTRACE_SIZE 3\n"
+        "#define BACKTRACE_ENTRY 1\n"
         "\n"
-        "static inline void QMPI_Get_function(int calling_tool_id, enum QMPI_Functions_enum function_enum,\n"
+        "static inline int QMPI_Get_function(int calling_tool_id, enum QMPI_Functions_enum function_enum,\n"
         "                      void (**function_ptr) (void), QMPI_Context *next_tool_context,\n"
         "                      int *next_tool_id)\n"
         "{\n"
         "    int mpi_errno = MPI_SUCCESS;\n"
         "    QMPI_Context context;\n"
         "\n"
+        "    void *array[BACKTRACE_SIZE];\n"
+        "    char **traces;\n"
+        "    int size, i;\n"
+        "    char *target = NULL;\n"
+        "\n"
         "    context.storage_stack = MPIR_QMPI_storage;\n"
         "\n"
+        "    // Capture the calling function address\n"
+        "    size = backtrace(array, BACKTRACE_SIZE);\n"
+        "    traces = backtrace_symbols(array, size);\n"
+        "    if (traces != NULL) {\n"
+        "        // Extract the address from the backtrace\n"
+        "        if (size > BACKTRACE_ENTRY) {\n"
+        "            const char *start_pattern = \"[\";\n"
+        "            const char *end_pattern = \"]\";\n"
+        "\n"
+        "            char *start, *end;\n"
+        "            if (start = strstr(traces[BACKTRACE_ENTRY], start_pattern)) {\n"
+        "                start += strlen(start_pattern);\n"
+        "                if (end = strstr(start, end_pattern)) {\n"
+        "                    target = (char *) malloc(end - start + 1);\n"
+        "                    memcpy(target, start, end - start);\n"
+        "                    target[end - start] = '\\0';\n"
+        "                }\n"
+        "            }\n"
+        "\n"
+        "        }\n"
+        "    }\n"
+        "    free(traces);\n"
+        "    // Store the calling address in the current context\n"
+        "    context.calling_address = calloc(0, strlen(target) + 1);\n"
+        "    memcpy(context.calling_address, target, strlen(target) + 1);\n"
+        "    free(target);\n"
+        "\n"
```

Comment:
Does this do anything to detect whether there is a PMPI tool attached and skip that one?

Reply:
No, currently, it does not. I will add that logic to skip.

Reply:
Added the logic to look for PMPI_ calls.